File: MultiLanguageSandbox/src/test/deps/pascal/31.pas
{$ASSERTIONS ON}

uses
  SysUtils;
(*
  Function: minRoundsToSameChar
  Given a string consisting of lowercase English letters, this function
  calculates the minimum number of rounds needed to modify the string so
  that all of its characters are the same. In each round, you can change
  one character to another character.

  Examples:
    - minRoundsToSameChar('aab') -> 1
*)

function minRoundsToSameChar(s: AnsiString): Integer;
var
  freq: array['a'..'z'] of Integer;
  i, maxFreq: Integer;
begin
  // Initialize frequency array
  for i := Ord('a') to Ord('z') do
    freq[Chr(i)] := 0;

  // Count the frequency of each character in the string
  for i := 1 to Length(s) do
    Inc(freq[s[i]]);

  // Find the maximum frequency among all characters
  maxFreq := 0;
  for i := Ord('a') to Ord('z') do
  begin
    if freq[Chr(i)] > maxFreq then
      maxFreq := freq[Chr(i)];
  end;

  // The result is the total length minus the maximum frequency
  minRoundsToSameChar := Length(s) - maxFreq;
end;
procedure testMinRoundsToSameChar;
begin
  Assert(minRoundsToSameChar('aab') = 1);
  Assert(minRoundsToSameChar('abc') = 2);
  Assert(minRoundsToSameChar('aaa') = 0);
  Assert(minRoundsToSameChar('abab') = 1);
  Assert(minRoundsToSameChar('zzzzz') = 0);
end;

begin
  testMinRoundsToSameChar;
end.

File: MultiLanguageSandbox/src/test/deps/javascript/31.js

/**
 * Given a string consisting of lowercase English letters, you can change one character to another
 * in each round. This function calculates the minimum number of rounds needed to make the string
 * composed of the same character.
 * @param {string} s - The input string.
 * @returns {number} The minimum number of rounds to make all characters in the string the same.
 *
 * Examples:
 * minRoundsToSameChar("aab") // returns 1
 * minRoundsToSameChar("abc") // returns 2
 * minRoundsToSameChar("aaa") // returns 0
 */

function minRoundsToSameChar(s) {
    const frequency = {};
    let maxFrequency = 0;
    
    for (const char of s) {
        if (frequency[char]) {
            frequency[char]++;
        } else {
            frequency[char] = 1;
        }
        if (frequency[char] > maxFrequency) {
            maxFrequency = frequency[char];
        }
    }
    
    return s.length - maxFrequency;
}
const testMinRoundsToSameChar = () => {
    console.assert(minRoundsToSameChar("aab") === 1, "Expected 1 round for 'aab'");
    console.assert(minRoundsToSameChar("abc") === 2, "Expected 2 rounds for 'abc'");
    console.assert(minRoundsToSameChar("aaa") === 0, "Expected 0 rounds for 'aaa'");
    console.assert(minRoundsToSameChar("abab") === 1, "Expected 1 round for 'abab'");
    console.assert(minRoundsToSameChar("zzzzz") === 0, "Expected 0 rounds for 'zzzzz'");
};

testMinRoundsToSameChar();